task_geo/data_sources/covid_cases/all_countries/all_countries.py

```diff
@@ -0,0 +1,78 @@
+import pandas as pd
+import numpy as np
+import time
+
+
+def all_country_datasource_connector(*args, **kwargs):
+  '''
+    Input:  None
+    Output: Return raw dataframe of all countries
+  '''
+  df = pd.read_csv("https://coronadatascraper.com/timeseries.csv")
+  
+  return df
+
+def all_country_datasource_formatter(df):
+  '''
+    Input:  Raw data from connector method
+    Output: Return list of dataframe of all countries
+  '''
+
+  df_list = []
+
+  #Remove rows where country is not provided
+  df = df[df['country'].notna()]
+  #df = df[df['state'].notna()]
+
+  #Get all country list present in dataset
+
+  country_list = df.country.unique()
+
+
+  #Currently all Country are having null state data values. 
+  #Below code will extract Geo-Data as country wise and then save it.
+
+  for country in country_list:
+    
+    idx = 0
+
+    #Create empty dataframe for storing Country-wise data from main dataframe
+    fnl = pd.DataFrame(columns=['state','country','population','lat','long','url','aggregate','tz','date','cases','deaths','recovered','active','growthFactor'])
+
+    ed = df[df['country']==country].sort_values(by='date')
+
+    #Store values ony by one for each country sorted by date
+    for date in ed.date.unique():
+      fnl.loc[idx] = pd.Series()
+      fnl.loc[idx]['state'] = ed[ed['date']==date]['state'].values[0]
+      fnl.loc[idx]['country'] = ed[ed['date']==date]['country'].values[0]
+      fnl.loc[idx]['population'] = ed[ed['date']==date]['population'].values[0]
+      fnl.loc[idx]['lat'] = ed[ed['date']==date]['lat'].values[0]
+      fnl.loc[idx]['long'] = ed[ed['date']==date]['long'].values[0]
+      fnl.loc[idx]['url'] = ed[ed['date']==date]['url'].values[0]
+      fnl.loc[idx]['aggregate'] = ed[ed['date']==date]['aggregate'].values[0]
+      fnl.loc[idx]['tz'] = ed[ed['date']==date]['tz'].values[0]
+      fnl.loc[idx]['date'] = date
+      fnl.loc[idx]['cases'] = ed[ed['date']==date]['cases'].values[0]
+      fnl.loc[idx]['deaths'] = ed[ed['date']==date]['deaths'].values[0]
+      fnl.loc[idx]['recovered'] = ed[ed['date']==date]['recovered'].values[0]
+      fnl.loc[idx]['active'] = ed[ed['date']==date]['active'].values[0]
+      fnl.loc[idx]['growthFactor'] = ed[ed['date']==date]['growthFactor'].values[0]
+      idx += 1
```

Comment:
This is wrong on various levels:

1. As a general rule: 
   a) Don't write code that can be generate with a for loop, 
   b) Use enumerate instead of a counter like `idx` that you increment on each loop.
So, all these lines could be condensed to:
```python
for (idx, date) in enumerate(ed.date.unique()):
   fnl = pd.Series()
   for column in columns:
       fnl.loc[idx][column] = ed[ed['date']==date][column].values[0]
```
2. As pandas rules:
    a) Never iterate(using vanilla Python) through a `pandas.DataFrame` to generate another `pandas.DataFrame`. **Never**. If you found yourself doing this you can be sure that something is wrong.
    b) Be aware with repeated filtering instead of doing: `ed[ed[' date' ] == date]` on each line or on each iteration of the for loop, do it **once** before entering the loop, store it in another variable an use it, I mean:
```python 
dated = ed[ed[' date' ] == date]
for column in columns: # Don't do this, is bad as I explained on the point a), this is only as example!
    fnl.loc[idx][column] = dated[column].values[0]
```
c) Assigning the values of the row one by one, is a bad practice, you can vectorize:
```python
for date in ed.date.unique():
    fnl.loc[idx] = ed[ed[' date' ] == date]
```

And the final remark about this lines is that all these lines can be removed, because `ed` as is generated on line 42 and `fnl` as is appended to `df_list` are **exactly the same**, you could simply append `ed` to the list, so this lines of code were redundant.

I just bother to remark all these things in an effort to help you become better with using pandas and a better developer.